Mathlib/Combinatorics/SimpleGraph/Walk.lean

```diff
@@ -765,61 +770,81 @@ lemma notNilRec_cons {motive : {u w : V} → (p : G.Walk u w) → ¬ p.Nil → S
     motive (q.cons h) Walk.not_nil_cons) (h' : G.Adj u v) (q' : G.Walk v w) :
     @Walk.notNilRec _ _ _ _ _ cons _ _ = cons h' q' := by rfl
 
-/-- The second vertex along a non-nil walk. -/
-def sndOfNotNil (p : G.Walk v w) (hp : ¬ p.Nil) : V :=
-  p.notNilRec (@fun _ u _ _ _ => u) hp
-
-@[simp] lemma adj_sndOfNotNil {p : G.Walk v w} (hp : ¬ p.Nil) :
-    G.Adj v (p.sndOfNotNil hp) :=
-  p.notNilRec (fun h _ => h) hp
+@[simp] lemma adj_getVert_one {p : G.Walk v w} (hp : ¬ p.Nil) :
+    G.Adj v (p.getVert 1) := by
+  have := adj_getVert_succ p (by
+    rw [SimpleGraph.Walk.nil_iff_length_eq] at hp
+    push_neg at hp
+    exact Nat.zero_lt_of_ne_zero hp)
+  rwa [getVert_zero] at this
+
+/-- The walk obtained by removing the first n darts of a walk. -/
+def drop {u v : V} (p : G.Walk u v) (n : ℕ) : G.Walk (p.getVert n) v :=
+  match p, n with
+  | .nil, _ => .nil
+  | p, 0 => p.copy (getVert_zero p).symm rfl
+  | .cons h q, (n + 1) => (q.drop n).copy (getVert_cons_succ _ h).symm rfl
 
 /-- The walk obtained by removing the first dart of a non-nil walk. -/
-def tail (p : G.Walk u v) (hp : ¬ p.Nil) : G.Walk (p.sndOfNotNil hp) v :=
-  p.notNilRec (fun _ q => q) hp
+def tail (p : G.Walk u v) : G.Walk (p.getVert 1) v := p.drop 1
```

Comment:
Yeah now it kind of looks like we could just remove this definition entirely. I need to think more...

Reply:
My thinking here is tailing a walk is a relatively common operation, dropping n not so much. I think from a generic software engineering perspective I think it makes sense to have tail because it clearly communicates what you're doing and it also intuitively matches what you expect from tail in a functional context.

Reply:
I agree. I guess we should a bit longer to see whether more uses appear in Mathlib

Reply:
Just to be clear, I was defending keeping tail, but are you proposing removing drop? (I'm actually tempted to keep drop too, since it's a piece of future engineering that has negligent maintenance burden, we could add drop lemma's when drop is actually used)

Reply:
I was proposing removing tail. To be clear, no action needs to be taken in this PR, and I tend to agree with your usability argument.